core/src/main/scala/cats/Parallel.scala

```diff
@@ -0,0 +1,220 @@
+package cats
+
+import cats.arrow.FunctionK
+
+/**
+  * Some types that form a FlatMap, are also capable of forming an Apply that supports parallel composition.
+  * The NonEmptyParallel type class allows us to represent this relationship.
+  */
+trait NonEmptyParallel[M[_], F[_]] extends Serializable {
+  /**
+    * The Apply instance for F[_]
+    */
+  def applicative: Apply[F]
+
+  /**
+    * The FlatMap instance for M[_]
+    */
+  def monad: FlatMap[M]
+
+  /**
+    * Natural Transformation from the parallel Apply F[_] to the sequential FlatMap M[_].
+    */
+  def sequential: F ~> M
+
+  /**
+    * Natural Transformation from the sequential FlatMap M[_] to the parallel Apply F[_].
+    */
+  def parallel: M ~> F
+
+}
+
+/**
+  * Some types that form a Monad, are also capable of forming an Applicative that supports parallel composition.
+  * The Parallel type class allows us to represent this relationship.
+  */
+trait Parallel[M[_], F[_]] extends NonEmptyParallel[M, F] {
+  /**
+    * The applicative instance for F[_]
+    */
+  def applicative: Applicative[F]
+
+  /**
+    * The monad instance for M[_]
+    */
+  def monad: Monad[M]
+
+  /**
+    * Natural Transformation from the parallel Applicative F[_] to the sequential Monad M[_].
+    */
+  def sequential: F ~> M
+
+  /**
+    * Natural Transformation from the sequential Monad M[_] to the parallel Applicative F[_].
+    */
+  def parallel: M ~> F
+
+  /**
+    * Provides an `ApplicativeError[F, E]` instance for any F, that has a `Parallel[M, F]`
+    * and a `MonadError[M, E]` instance.
+    * I.e. if you have a type M[_], that supports parallel composition through type F[_],
+    * then you can get `ApplicativeError[F, E]` from `MonadError[M, E]`.
+    */
+  def applicativeError[E](implicit E: MonadError[M, E]): ApplicativeError[F, E] = new ApplicativeError[F, E] {
+
+    def raiseError[A](e: E): F[A] =
+      parallel(MonadError[M, E].raiseError(e))
+
+    def handleErrorWith[A](fa: F[A])(f: (E) => F[A]): F[A] = {
+      val ma = MonadError[M, E].handleErrorWith(sequential(fa))(f andThen sequential.apply)
+      parallel(ma)
+    }
+
+    def pure[A](x: A): F[A] = applicative.pure(x)
+
+    def ap[A, B](ff: F[(A) => B])(fa: F[A]): F[B] = applicative.ap(ff)(fa)
```

Comment:
I think we should override basically everything here or we could de-optimize an existing `applicative`.